### PR TITLE
Server initializers

### PR DESCRIPTION
**Pull request by antonmes**
This pr provides more natural widely used api to create and run the Server

### CODE REVIEW

**Comment by dreymonde** [comment]
Personally, I like this more than `server.serve`

**Comment by paulofaria** [comment]
The reason I prefer `serve` is because that's the real action servers perform. It's *the* verb that describes what they do, they *serve* apps on a give host and port. Also, enforcing an initializer, might have it's advantages. Like you can have an API that receives `Server.Type` and then you initialize the server yourself. I can see a big advantage in that. This can make dependency injection really easy. But in this case I would prefer it like this:

```swift
public protocol Server {
    init()
    func serve(_ responder: Responder, on host: String, at port: Int) throws
}
```

Another reason why this makes sense. Using Zewo's server as an example. You can initialize Zewo's server passing a lot of options:

```swift
let app = Router { route in
    route.get("/") { _ in
        return Response(body: "Hello")
    }
}

let server = Server(reusingPort: true, parsingWith: FastParser.Type)
server.serveInBackground(app, at: 80)
server.serve(app, at: 443, certificate: "server.crt", privateKey: "server.key")
```

**Comment by antonmes** [comment]
I can't see advantages of `serve` function, same think can be done thought initializer but it more natural way. Also in case of ssl keys i think it's better to pass them as struct like SSLConfiguration() or SSLKeys() to initializer or public property like `server.SSLConfiguration = SSLConfiguration()`, because we also will have Proxy Credentials and other configurations, and `serve` is not helping with this at all.

As for the `serve` meaning, i don't think it really `serves` the app: if it is, it should do all the work in Server class and let Router and other classes to be just a xml configuration or something like that, but it isn't, it just starts listening socket, and passes accepted requests to the `handler` or `responder` in our case.

And the last thing, if OpenSwift is going to be a standard, it shouldn't neither surprise nor confuse peoples. And `serve` confrise me a lot. 

https://swift.org/documentation/api-design-guidelines/

> - Stick to the established meaning if you do use a term of art.
> 
>   The only reason to use a technical term rather than a more common word is that it precisely expresses something that would otherwise be ambiguous or unclear. Therefore, an API should use the term strictly in accordance with its accepted meaning.
> 
>   
> 
>   - Don’t surprise an expert: anyone already familiar with the term will be surprised and probably angered if we appear to have invented a new meaning for it.
> 
>   - Don’t confuse a beginner: anyone trying to learn the term is likely to do a web search and find its traditional meaning.

File: Sources/AsyncServer.swift [hunk 1]
```diff
@@ -1,9 +1,10 @@
 public protocol AsyncServer {
-    func serve(responder: AsyncResponder, on host: String, at port: Int) throws
+    init(host: String, port: Int, responder: AsyncResponder) throws
+    func start() throws
 }
 
 extension AsyncServer {
-    public func serve(responder: AsyncResponder, at port: Int) throws {
-        try self.serve(responder, on: "0.0.0.0", at: port)
+    public init(port: Int, responder: AsyncResponder) throws {
```
Comment (scottbyrns) [inline Sources/AsyncServer.swift]: While I'm not a fan of protocols that force initializers it does make sense.

LGTM

**Comment by Danappelxx** [comment]
LGTM, though I'm a little concerned that doing it this way might make it harder to inject servers. 

cc @tannernelson

**Comment by paulofaria** [comment]
LGTM

**Comment by antonmes** [comment]
testing slack bot

**Comment by tannernelson** [comment]
I'm worried forcing this initializer will make it harder for classes to conform (ie they will need to have optional member variables). 

Unless this is breaking something I don't think we should change it.

**Comment by paulofaria** [comment]
@tannernelson They won't need optional. They can have defaults, which they should anyway.

**Comment by paulofaria** [comment]
Also having the initializers are great actually. Cause then a given API can initialize the server itself. So you can just give it the type and it does it all automatically.

**Comment by Danappelxx** [comment]
Actually, I think allowing the 'given API' to initializer the server itself entirely defeats the purpose, doesn't it? We want to _inject_ the server into the api (hence the term dependency injection). I was for this at first but now I don't think its a good idea.

**Comment by paulofaria** [comment]
Having the `init` allows more freedom. You can still inject the dependency initializing it before passing it. But with the initializer a given API can take control over the parameters that are sent to the Server. Of course the parameters would be limited and the rest would have to be defaulted, but this is still better than not being able to do it at all.

In other words. Having the `init` doesn't restrain the `Server` in any way. It only allows more freedom.

**Comment by paulofaria** [comment]
Also, injecting the `Type` you want is still dependency injection. Actually that's the most important information in dependency injection, the `Type`. The parameters are very important, of course. But what really counts in dependency injection is the `Type`.

**Comment by Danappelxx** [comment]
No, you're wrong. Passing the type isn't enough for dependency injection. With this pr, how would you pass in HTTPSServer instead of HTTPServer? You can't, because the app (ex vapor) would have to know about your certificates, which it doesn't and shouldn't.

**Comment by antonmes** [comment]
@Danappelxx what the difference between passing `Responder` to `init` or `func`? You should now about HTTPSServer certificates in both cases. In my opinion there is no restrictions or freedom in choosing one, they are the same, but `init` is more `standard` and `familiar` api. 
Could you write an example, what we _can_ do with current api and _can't_ with the new one?

**Comment by Danappelxx** [comment]
Right, here's an example:
```swift
let server = HTTPSServer(reusePort: true, certificates:...)
app.use(server)
app.start() 
// we can now have multiple instances sharing the port, along with having a configured ssl
```
vs
```swift
app.use(HTTPSServer.self)
app.start()
// we cannot start multiple instances since they don't share the port by default, and ssl is not configured
```

In the second example, the app needs to have special knowledge about its http server, otherwise its impossible to configure it. Unless I'm missing something, this approach effectively makes dependency injection useless.

**Comment by antonmes** [comment]
I think we are talking about different pull requests 😄

this one is about
```swift
let server = HTTPServer()
server.serve(host: "0.0.0.0", port: 8080)
```
vs 
```swift
let server = HTTPServer(host: "0.0.0.0", port: 8080)
server.run()
```
thats it

**Comment by Danappelxx** [comment]
No, actually, we're talking about the same one 😃 (had me triple-check my sanity for a second there). Since the initializer is defined, the _type_ is meant to be injected, not an instance. There is little difference outside of that, but I think its very important that we keep dependency injection as powerful as possible. I showed in my previous comment what the consequences of this pr are in that context.

**Comment by antonmes** [comment]
@Danappelxx but you still have to implement `serve(host:port:)`? so what the difference?
Could you show another example?

**Comment by Danappelxx** [comment]
I don't think you understand. Initializing the server yourself gives you an opportunity to configure it before you pass it on to the app. If you let the app instantiate it, you cannot change anything about the server except what the app knows about it. The whole idea of polymorphism is that you just have to meet the requirements (which is in this case is just `server(host:port:)`), but everything else can be modified as you choose.

**Comment by antonmes** [comment]
@Danappelxx 
what polymorphism? where did you get `HTTPSServer(reusePort: true, certificates:...)` then? it's not a protocol right? So why can't you use: 
```swift
let server = HTTPSServer(host: "0.0.0.0", port: 8080, reusePort: true, certificates:...)
app.use(server)
app.start() 
```

**Comment by paulofaria** [comment]
@dan having the initializer still allows you to initialize the server however you want and pass it already initialized to whoever needs it. it doesn't restrain in any way. 

```
let server = HTTPSServer(reusePort: true, certificates:...)
app.use(server)
app.start() 
// we can now have multiple instances sharing the port, along with having a configured ssl
```

vs

```
app.use(HTTPSServer.self)
app.start()
// we cannot start multiple instances since they don't share the port by default, and ssl is not configured
```

the decision of receiving a pre-initialized injection or a type injection is up to the API receiving the server so we should allow this.

**Comment by paulofaria** [comment]
Bottomline. Having the initializer allows more freedom without any further restraining, so we should provide it. And I would say we should add canonical initializers not only to `Server`, but to other protocols, for the same reason. But this should be done in another PR.

**Comment by tannernelson** [comment]
My confusion here was thinking that initializing the server also caused it to start listening. If you still need to call `.run()` or something similar to cause the server to bind to the given port, then I'm behind this PR 👍

**Comment by antonmes** [comment]
2 weeks, no objections, merging.

**Comment by Danappelxx** [comment]
Well, I'm actually still against this. Even if it allows the user to configure it, it still restricts the configuration in some ways. For example, if I inject an instance of HTTPServer into my application and then want to set the port from some environmental variable (from inside the application code), I can no longer do that. I believe that alone is pretty big and would actually break the Zewo+Vapor bridge.

Seriously, if we care about proper dependency injection and configuration, this pr is a pretty big step backwards.

**Comment by antonmes** [comment]
>For example, if I inject an instance of HTTPServer into my application and then want to set the port from some environmental variable (from inside the application code), I can no longer do that

But before, we couldn't set the port when instantiating HTTPServer `¯\_(ツ)_/¯` just vice versa
There is no restrictions on specifying the port from some environmental variable, you still can do that.

**Comment by Danappelxx** [comment]
>There is no restrictions on specifying the port from some environmental variable, you still can do that.

Right, but maybe you dont get it from an environmental variable and instead through a cli or configuration file. The app  _needs_ to have basic access to configuration. 

What about adding port and host property requirements to this protocol (with both get and set defined)? That would actually fix most of my complaints.

**Comment by paulofaria** [comment]
@Danappelxx It's not a step back, it's a step forward. 

> For example, if I inject an instance of HTTPServer into my application and then want to set the port from some environmental variable (from inside the application code), I can no longer do that

Yeah this is something we can add that would improve even further dependency injection. @Danappelxx Can you send a PR?

**Comment by paulofaria** [comment]
Actually I'll send it now.

**Comment by Danappelxx** [comment]
I think you quoted the wrong comment, but yeah, that would be great!

**Pull request merged**